### PR TITLE
refactor transform lib

### PR DESCRIPTION
**Pull request by yye-lyft**
Refactor the reference system such that targets are now in a new space called `agent` and not in `world` anymore.

### CODE REVIEW

File: l5kit/l5kit/geometry/transform.py [hunk 1]
```diff
@@ -33,55 +33,6 @@ def yaw_as_rotation33(yaw: float) -> np.ndarray:
     return transforms3d.euler.euler2mat(0, 0, yaw)
 
 
-def world_to_image_pixels_matrix(
```
Comment (lucabergamini) [inline l5kit/l5kit/geometry/transform.py:36]: Please can we have this back with the new code to compute it? You have 6 lines of code repeated in 4 places to compute this. It's valuable to have a function to do it

Comment (lucabergamini) [reply l5kit/l5kit/geometry/transform.py:36]:
```python
def raster_from_world(render_context: RenderContext, ego_translation_m: np.ndarray, ego_yaw_rad: float) -> np.ndarray:
    """
    Return a matrix to convert a world pose into raster coordinates

    Args:
        render_context (RenderContext): the context for rasterisation
        ego_translation_m (np.ndarray): XY translation in world coordinates
        ego_yaw_rad (float): yaw angle

    Returns:
        (np.ndarray): 3x3 matrix
    """
    # Compute ego pose from its position and heading
    ego_pose = np.array([[np.cos(ego_yaw_rad), -np.sin(ego_yaw_rad), ego_translation_m[0]],
                         [np.sin(ego_yaw_rad), np.cos(ego_yaw_rad), ego_translation_m[1]],
                         [0, 0, 1]])

    ego_from_world = np.linalg.inv(ego_pose)
    return render_context.raster_from_local @ ego_from_world
```

File: l5kit/l5kit/rasterization/render_context.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import numpy as np
+
+
+class RenderContext:
+    def __init__(
+        self, raster_size_px: np.ndarray, pixel_size_m: np.ndarray, center_in_raster_ratio: np.ndarray,
+    ) -> None:
+        """
+        Arguments:
```
Comment (lucabergamini) [inline l5kit/l5kit/rasterization/render_context.py:9]: Can we add a description of what this does?

File: l5kit/l5kit/rasterization/render_context.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        if pixel_size_m[0] != pixel_size_m[1]:
+            raise NotImplementedError("No support for non squared pixels yet")
+
+        self.raster_size_px = raster_size_px
+        self.pixel_size_m = pixel_size_m
+        self.center_in_raster_ratio = center_in_raster_ratio
+
+        scaling = 1.0 / pixel_size_m  # scaling factor from world to raster space [pixels per meter]
+        center_in_raster_px = center_in_raster_ratio * raster_size_px
+        self.raster_from_local = np.array([[scaling[0], 0, center_in_raster_px[0]],
```
Comment (lucabergamini) [inline l5kit/l5kit/rasterization/render_context.py:25]: this is the intrinsic matrix in fact

File: l5kit/l5kit/rasterization/render_context.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+                                           [0, scaling[1], center_in_raster_px[1]],
+                                           [0, 0, 1]])
+
+    def raster_from_local(self) -> np.ndarray:
```
Comment (lucabergamini) [inline l5kit/l5kit/rasterization/render_context.py:29]: any reason not to use the variable itself?

File: l5kit/l5kit/rasterization/sat_box_rasterizer.py [hunk 5]
```diff
@@ -34,9 +36,11 @@ def __init__(
         self.world_to_aerial = world_to_aerial
         self.interpolation = interpolation
 
-        self.box_rast = BoxRasterizer(raster_size, pixel_size, ego_center, filter_agents_threshold, history_num_frames)
+        self.box_rast = BoxRasterizer(
```
Comment (lucabergamini) [inline l5kit/l5kit/rasterization/sat_box_rasterizer.py:39]: don't you already have some of these information in the `render_context`?

File: l5kit/l5kit/sampling/agent_sampling.py [hunk 6]
```diff
@@ -86,13 +83,10 @@ def generate_agent_sample(
     history_agents = filter_agents_by_frames(history_frames, agents)
     future_agents = filter_agents_by_frames(future_frames, agents)
 
-    try:
```
Comment (lucabergamini) [inline l5kit/l5kit/sampling/agent_sampling.py:89]: note, you're not on latest master

File: l5kit/l5kit/sampling/agent_sampling.py [hunk 7]
```diff
@@ -153,10 +147,10 @@ def generate_agent_sample(
         "history_positions": history_coords_offset,
         "history_yaws": history_yaws_offset,
         "history_availabilities": history_availability,
-        "world_to_image": world_to_image_space,
-        "centroid": agent_centroid,
-        "yaw": agent_yaw,
-        "extent": agent_extent,
+        "world_to_image": raster_from_world,
```
Comment (lucabergamini) [inline l5kit/l5kit/sampling/agent_sampling.py:150]: we should create a `raster_from_world` with the same value and start deprecating `world_to_image`

File: l5kit/l5kit/sampling/agent_sampling.py [hunk 8]
```diff
@@ -190,22 +184,22 @@ def _create_targets_for_deep_prediction(
...
         else:
             # it's not guaranteed the target will be in every frame
             try:
-                agent = filter_agents_by_track_id(agents, selected_track_id)[0]
+                agent = filter_agents_by_track_id(frame_agents, selected_track_id)[0]
             except IndexError:
                 availability[i] = 0.0  # keep track of invalid futures/history
                 continue
 
             agent_centroid = agent["centroid"]
```
Comment (lucabergamini) [inline l5kit/l5kit/sampling/agent_sampling.py:199]: we can move this into the `try` branch

File: l5kit/l5kit/sampling/agent_sampling.py [hunk 9]
```diff
@@ -190,22 +184,22 @@ def _create_targets_for_deep_prediction(
...
             agent_yaw = agent["yaw"]
 
-        coords_offset[i] = agent_centroid - agent_current_centroid
-        yaws_offset[i] = angular_distance(agent_yaw, agent_current_yaw)
+        coords_offset[i] = transform_points(agent_centroid, current_agent_from_world)
```
Comment (lucabergamini) [inline l5kit/l5kit/sampling/agent_sampling.py:202]: interesting, local and local displacement are the same now (makes sense)

**Review by lucabergamini — COMMENTED** [review]
please restore transform function.
some minors

File: l5kit/l5kit/geometry/transform.py (hunk 1 above)
Comment (yye-lyft) [reply l5kit/l5kit/geometry/transform.py:36]: make sense

File: l5kit/l5kit/rasterization/render_context.py (hunk 3 above)
Comment (yye-lyft) [reply l5kit/l5kit/rasterization/render_context.py:25]: Yes

File: l5kit/l5kit/rasterization/sat_box_rasterizer.py (hunk 5 above)
Comment (yye-lyft) [reply l5kit/l5kit/rasterization/sat_box_rasterizer.py:39]: you mean `raster_size`?

File: l5kit/l5kit/sampling/agent_sampling.py (hunk 8 above)
Comment (yye-lyft) [reply l5kit/l5kit/sampling/agent_sampling.py:199]: I rebased to latest master

File: l5kit/l5kit/geometry/transform.py (hunk 1 above)
Comment (yye-lyft) [reply l5kit/l5kit/geometry/transform.py:36]: I moved `raster_from_world` function to render context class

File: l5kit/l5kit/rasterization/sat_box_rasterizer.py (hunk 5 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/sat_box_rasterizer.py:39]: isn't:
 `raster_size_px` == `raster_size`
`pixel_size_m` == `pixel_size`
`center_in_raster_ratio` == `ego_center`
?

File: l5kit/l5kit/geometry/transform.py [hunk 10]
```diff
@@ -33,55 +54,6 @@ def yaw_as_rotation33(yaw: float) -> np.ndarray:
...
-
-    Returns:
-        np.ndarray: 3x3 transformation matrix
-    """
-
-    # Translate world to ego by applying the negative ego translation.
-    world_to_ego_in_2d = np.eye(3, dtype=np.float32)
-    world_to_ego_in_2d[0:2, 2] = -ego_translation_m[0:2]
-
-    if ego_yaw_rad is not None:
```
Comment (yye-lyft) [inline l5kit/l5kit/geometry/transform.py:65]: @lucabergamini does it happen? Shall we add check in our current codebase?

File: l5kit/l5kit/rasterization/render_context.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            raise NotImplementedError("No support for non squared pixels yet")
+
+        self.raster_size_px = raster_size_px
+        self.pixel_size_m = pixel_size_m
+        self.center_in_raster_ratio = center_in_raster_ratio
+
+        scaling = 1.0 / pixel_size_m  # scaling factor from world to raster space [pixels per meter]
+        center_in_raster_px = center_in_raster_ratio * raster_size_px
+        self.raster_from_local = np.array(
+            [[scaling[0], 0, center_in_raster_px[0]], [0, scaling[1], center_in_raster_px[1]], [0, 0, 1]]
```
Comment (yye-lyft) [inline l5kit/l5kit/rasterization/render_context.py:29]: Can we recover the 3x3 matrix? This flat matrix is not as clear as before

File: l5kit/l5kit/tests/evaluation/extract_metric_test.py [hunk 12]
```diff
@@ -7,11 +7,12 @@
 from l5kit.dataset import AgentDataset
 from l5kit.evaluation import compute_metrics_csv, write_gt_csv, write_pred_csv
 from l5kit.evaluation.metrics import neg_multi_log_likelihood
-from l5kit.rasterization import StubRasterizer
+from l5kit.rasterization import RenderContext, StubRasterizer
 
 
 def test_compute_mse_error(tmp_path: Path, zarr_dataset: ChunkedDataset, cfg: dict) -> None:
-    rast = StubRasterizer((10, 10), np.asarray((0.25, 0.25)), np.asarray((0.5, 0.5)), 0.5)
+    context = RenderContext(np.asarray((10, 10)), np.asarray((0.25, 0.25)), np.asarray((0.5, 0.5)))
```
Comment (yye-lyft) [inline l5kit/l5kit/tests/evaluation/extract_metric_test.py:14]: `context`->`render_context`

File: l5kit/l5kit/tests/geometry/transform_test.py [hunk 13]
```diff
@@ -2,34 +2,43 @@
 import pytest
 import transforms3d
 
-from l5kit.geometry import transform_point, transform_points, transform_points_transposed, world_to_image_pixels_matrix
+from l5kit.geometry import transform_point, transform_points, transform_points_transposed
+from l5kit.rasterization.render_context import RenderContext
 
 
 def test_transform_to_image_space_2d() -> None:
```
Comment (yye-lyft) [inline l5kit/l5kit/tests/geometry/transform_test.py:9]: I would move this unit test to `l5kit/tests/rasterization/render_context_test.py`

File: l5kit/l5kit/rasterization/render_context.py (hunk 11 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/render_context.py:29]: I'll try, but it's `black` here I think

File: l5kit/l5kit/sampling/agent_sampling.py [hunk 14]
```diff
@@ -149,10 +142,13 @@ def generate_agent_sample(
...
-        "centroid": agent_centroid,
-        "yaw": agent_yaw,
-        "extent": agent_extent,
+        "world_to_image": raster_from_world,  # TODO deprecate
+        "raster_from_agent": raster_from_world @ np.linalg.inv(agent_from_world),
+        "raster_from_world": raster_from_world,
+        "agent_from_world": agent_from_world,
+        "centroid": agent_centroid_m,
+        "yaw": agent_yaw_rad,
+        "extent": agent_extent_m,
```
Comment (yye-lyft) [inline l5kit/l5kit/sampling/agent_sampling.py:151]: After the challenge we should consider change key to `agent_position_m`, `agent_yaw_rad`, `agent_extent_m`

File: l5kit/l5kit/geometry/transform.py (hunk 10 above)
Comment (lucabergamini) [reply l5kit/l5kit/geometry/transform.py:65]: this code is not here anymore :)

File: l5kit/l5kit/rasterization/render_context.py (hunk 11 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/render_context.py:29]: black really want to reformat this xD

Comment (yye-lyft) [reply l5kit/l5kit/rasterization/render_context.py:29]: Try:

Comment (lucabergamini) [reply l5kit/l5kit/rasterization/render_context.py:29]: I know, but is it worth it :)? If you think it is I'm fine in doing it :)

Comment (yye-lyft) [reply l5kit/l5kit/rasterization/render_context.py:29]: yeah I know, leave it then.

**Review by yye-lyft — COMMENTED** [review]
I leave some comment, will take a look at the straight road unit test later

File: l5kit/l5kit/tests/geometry/transform_test.py [hunk 15]
```diff
@@ -2,34 +2,24 @@
...
-    tf = world_to_image_pixels_matrix(image_shape, pixel_size, offset)
-    output_points = transform_points(input_points, tf)
-
-    np.testing.assert_array_equal(output_points, expected_output_points)
+from l5kit.geometry import transform_point, transform_points, transform_points_transposed
+from l5kit.rasterization.render_context import RenderContext
 
 
 def test_transform_single_point() -> None:
+    image_shape = np.asarray((200, 200))
```
[the remark this answers is not in the document]
Comment (lucabergamini) [reply l5kit/l5kit/tests/geometry/transform_test.py:10]: I think I will have two functions here to test transform using hard-coded matrices and one function there to check the `raster_from_world` then

File: l5kit/l5kit/tests/rasterization/render_context_test.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
+import numpy as np
+
+from l5kit.geometry import transform_points
+from l5kit.rasterization.render_context import RenderContext
+
+
+def test_transform_to_image_space_2d() -> None:
+    image_shape = np.asarray((200, 200))
```
[the remark this answers is not in the document]
Comment (lucabergamini) [reply l5kit/l5kit/tests/rasterization/render_context_test.py:8]: FYI: just put a comment for things like this. The variable has some usages in the function so if I commit this I still need to rename the usage in a following commit.

File: coords_systems.md [hunk 17]
```diff
@@ -25,12 +25,28 @@ is expressed via a single [yaw angle](https://en.wikipedia.org/wiki/Yaw_(rotatio
 
 The origin of the **world** coordinate system is located at [[37°25'45.6"N, 122°09'15.7"W]](https://www.google.com/maps/place/37%C2%B025'45.6%22N+122%C2%B009'15.7%22W/@37.4293427,-122.1565407) in Palo Alto (California, USA).
 
+
+## Agent Coordinate System
+A common feature of the BEV's rasterisation is that the agent of interest is always aligned in the same direction.
+In L5Kit this direction is left (i.e. the hood of the agent of interest always points to the left side of the image).
```
Comment (oliver-scheel) [inline coords_systems.md:31]: Isn't this direction RIGHT?

File: coords_systems.md [hunk 18]
```diff
@@ -25,12 +25,28 @@ is expressed via a single [yaw angle](https://en.wikipedia.org/wiki/Yaw_(rotatio
...
+
+This metric space is referred to as **agent** and has the following features:
+- The agent's positions is at (0, 0);
```
Comment (oliver-scheel) [inline coords_systems.md:34]: position

File: coords_systems.md [hunk 19]
```diff
@@ -25,12 +25,28 @@ is expressed via a single [yaw angle](https://en.wikipedia.org/wiki/Yaw_(rotatio
...
+- The agent's yaw is 0.
+
+If you're using one of our high-level dataset objects (either `EgoDataset` or `AgentDataset`) to generate samples, you can 
+access the agent-from-world matrix using the `agent_from_world` key on the returned dict.
```
Comment (oliver-scheel) [inline coords_systems.md:38]: of (or in) the returned dict

File: l5kit/l5kit/geometry/transform.py [hunk 20]
```diff
@@ -1,10 +1,31 @@
-from typing import Optional, Sequence, Tuple, Union, cast
+from typing import Sequence, Union, cast
 
 import numpy as np
 import pymap3d as pm
 import transforms3d
 
 
+def agent_pose(agent_centroid_m: np.ndarray, agent_yaw_rad: float) -> np.ndarray:
```
Comment (oliver-scheel) [inline l5kit/l5kit/geometry/transform.py:8]: Any reason why this is called agent_pose and not in style of 'world_to_image_pixels_matrix'?

File: coords_systems.md (hunk 17 above)
Comment (lucabergamini) [reply coords_systems.md:31]: hahah yes it is xD

File: l5kit/l5kit/geometry/transform.py (hunk 20 above)
Comment (lucabergamini) [reply l5kit/l5kit/geometry/transform.py:8]: this is at the same time the `agent_pose`  and the `world_from_agent` matrix, I couldn’t come up with a better name :( @yye-lyft ?

File: l5kit/l5kit/rasterization/render_context.py [hunk 21]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
+
+
+class RenderContext:
+    def __init__(
+        self, raster_size_px: np.ndarray, pixel_size_m: np.ndarray, center_in_raster_ratio: np.ndarray,
+    ) -> None:
+        """
+        This class stores render context information (raster size, pixel size, raster center / principle point) and
```
Comment (oliver-scheel) [inline l5kit/l5kit/rasterization/render_context.py:9]: Is 'render' actually just another word for 'rasterizer'? Does it make sense to introduce this, adding a bit of 'confusion' by multiple terms for the same?

File: l5kit/l5kit/rasterization/render_context.py [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+        it computes a transformation matrix (raster_from_local) to transform a local pose to the raster center
```
Comment (oliver-scheel) [inline l5kit/l5kit/rasterization/render_context.py:10]: 'to the raster center' sounds weird - this matrix defines a constant offset to the raster's center? Is it really a transformation matrix then, does it transform every point to this center?

File: l5kit/l5kit/rasterization/render_context.py (hunk 21 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/render_context.py:9]: We use this notation internally, so we're trying to be consistent here :)

File: l5kit/l5kit/rasterization/render_context.py (hunk 22 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/render_context.py:10]: I guess it's just means that the local centre ends up in the raster centre. Let me rephrase it a bit :)

File: l5kit/l5kit/rasterization/satellite_rasterizer.py [hunk 23]
```diff
@@ -51,35 +48,30 @@ def rasterize(
     ) -> np.ndarray:
 
         if agent is None:
-            ego_translation = history_frames[0]["ego_translation"]
+            ego_translation_m = history_frames[0]["ego_translation"]
             # Note 2: it looks like we are assuming that yaw in ecef == yaw in sat image
```
Comment (oliver-scheel) [inline l5kit/l5kit/rasterization/satellite_rasterizer.py:52]: Somewhat confusing note, are we not sure?

File: l5kit/l5kit/sampling/agent_sampling.py [hunk 24]
```diff
@@ -100,9 +98,9 @@ def generate_agent_sample(
     cur_agents = history_agents[0]
 
     if selected_track_id is None:
-        agent_centroid = cur_frame["ego_translation"][:2]
-        agent_yaw = rotation33_as_yaw(cur_frame["ego_rotation"])
-        agent_extent = np.asarray((EGO_EXTENT_LENGTH, EGO_EXTENT_WIDTH, EGO_EXTENT_HEIGHT))
+        agent_centroid_m = cur_frame["ego_translation"][:2]
+        agent_yaw_rad = rotation33_as_yaw(cur_frame["ego_rotation"])
```
Comment (oliver-scheel) [inline l5kit/l5kit/sampling/agent_sampling.py:102]: Seen it a couple times now until here: do we sometimes use degrees or why is _rad necessary?

File: l5kit/l5kit/visualization/utils.py [hunk 25]
```diff
@@ -7,13 +7,11 @@
 
 PREDICTED_POINTS_COLOR = (0, 255, 255)
 TARGET_POINTS_COLOR = (255, 0, 255)
-REFERENCE_TRAJECTORY_POINT_COLOR = (255, 255, 0)
+REFERENCE_TRAJ_COLOR = (255, 255, 0)
```
Comment (oliver-scheel) [inline l5kit/l5kit/visualization/utils.py:10]: Minor, but I find this a bit consisitent: It is predicted_POINTS_color and thus Id also prefer the old version actually.

**Review by oliver-scheel — APPROVED** [review]
Some minors, otherwise LGTM as far as I can tell, would prefer additional reviews though.

File: l5kit/l5kit/sampling/agent_sampling.py (hunk 24 above)
Comment (lucabergamini) [reply l5kit/l5kit/sampling/agent_sampling.py:102]: it's for clarity and easy documentation :)

File: l5kit/l5kit/visualization/utils.py (hunk 25 above)
Comment (lucabergamini) [reply l5kit/l5kit/visualization/utils.py:10]: It was changed by @yye-lyft, I'm fine with both as long as there is `COLOR` at the end :)

File: l5kit/l5kit/rasterization/satellite_rasterizer.py (hunk 23 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/satellite_rasterizer.py:52]: I'm not sure of anything in that rasterizer 😄

File: l5kit/l5kit/visualization/utils.py (hunk 25 above)
Comment (yye-lyft) [reply l5kit/l5kit/visualization/utils.py:10]: hmm, for me Trajectory is already means a sequence of waypoints, so I am remove `POINT`. Sorry for piggypack this change in this PR. I will create new PR for this.

File: l5kit/l5kit/geometry/transform.py (hunk 20 above)
Comment (yye-lyft) [reply l5kit/l5kit/geometry/transform.py:8]: `agent_pose_world` == `world_from_agent`
Maybe we can rename this function to `compute_agent_pose` if you want. @lucabergamini

File: l5kit/l5kit/rasterization/satellite_rasterizer.py (hunk 23 above)
Comment (yye-lyft) [reply l5kit/l5kit/rasterization/satellite_rasterizer.py:52]: We can have another round of refactoring in qualify week

File: l5kit/l5kit/rasterization/satellite_rasterizer.py [hunk 26]
```diff
@@ -51,35 +48,30 @@ def rasterize(
...
-            ego_yaw = rotation33_as_yaw(history_frames[0]["ego_rotation"])
+            ego_yaw_rad = rotation33_as_yaw(history_frames[0]["ego_rotation"])
+
         else:
-            ego_translation = np.append(agent["centroid"], history_frames[0]["ego_translation"][-1])
+            ego_translation_m = np.append(agent["centroid"], history_frames[0]["ego_translation"][-1])
             # Note 2: it looks like we are assuming that yaw in ecef == yaw in sat image
-            ego_yaw = agent["yaw"]
```
Comment (stefaniespeichert) [inline l5kit/l5kit/rasterization/satellite_rasterizer.py:60]: Maybe also only have this note once if it has to be here at all?

File: l5kit/l5kit/rasterization/satellite_rasterizer.py [hunk 27]
```diff
@@ -51,35 +48,30 @@ def rasterize(
...
-            self.raster_size,
-            self.pixel_size,
-            ego_translation_m=ego_translation,
-            ego_yaw_rad=ego_yaw,
-            ego_center_in_image_ratio=self.ego_center,
-        )
+        raster_from_world = self.render_context.raster_from_world(ego_translation_m, ego_yaw_rad)
+        world_from_raster = np.linalg.inv(raster_from_world)
 
-        # get the center of the images in meters using the inverse of the matrix,
```
Comment (stefaniespeichert) [inline l5kit/l5kit/rasterization/satellite_rasterizer.py:70]: I'm being very picky here so please ignore if ou think its unnecessary but sometimes you say 'meters' and sometimes 'metres' if there's an easy fix for this (like a control +F in pycharm) maybe make it uniform otherwise ignore

File: l5kit/l5kit/rasterization/satellite_rasterizer.py [hunk 28]
```diff
@@ -51,35 +48,30 @@ def rasterize(
...
-        world_translation = transform_point(center_pixel, np.linalg.inv(world_to_image_space))
-        sat_translation = transform_point(np.append(world_translation, ego_translation[2]), self.world_to_aerial)
+        # Transform raster center to satellite coordinates (consider also z here)
+        center_in_raster_px = np.asarray(self.raster_size) * (0.5, 0.5)
+        center_in_world_m = transform_point(center_in_raster_px, world_from_raster)
+        center_in_aerial_px = transform_point(np.append(center_in_world_m, ego_translation_m[2]), self.world_to_aerial)
 
         # Note 1: there is a negation here, unknown why this is necessary.
         # My best guess is because Y is flipped, maybe we can do this more elegantly.
         sat_im = get_sat_image_crop_scaled(
```
Comment (stefaniespeichert) [inline l5kit/l5kit/rasterization/satellite_rasterizer.py:70]: Note 1 comes after note 2.  You sure you want to leave this in?

**Review by oliver-scheel — APPROVED** [review]
LGTM

File: l5kit/l5kit/rasterization/semantic_rasterizer.py [hunk 29]
```diff
@@ -130,31 +127,30 @@ def rasterize(
...
     def render_semantic_map(
-        self, center_world: np.ndarray, world_to_image_space: np.ndarray, tl_faces: np.ndarray
+        self, center_in_world: np.ndarray, raster_from_world: np.ndarray, tl_faces: np.ndarray
     ) -> np.ndarray:
         """Renders the semantic map at given x,y coordinates.
 
         Args:
-            center_world (np.ndarray): XY of the image center in world ref system
-            world_to_image_space (np.ndarray):
+            raster_center_in_global (np.ndarray): XY of the image center in world ref system
```
Comment (yye-lyft) [inline l5kit/l5kit/rasterization/semantic_rasterizer.py:152]:
```suggestion
            center_in_world (np.ndarray): XY of the image center in world ref system
```

File: l5kit/l5kit/rasterization/semantic_rasterizer.py [hunk 30]
```diff
@@ -130,31 +127,30 @@ def rasterize(
         # TODO TR_FACES
```
Comment (stefaniespeichert) [inline l5kit/l5kit/rasterization/semantic_rasterizer.py:127]: Note the todo here! is it okay that this is still in? Also shouldn't it be TL_Faces?

File: l5kit/l5kit/rasterization/stub_rasterizer.py [hunk 31]
```diff
@@ -11,24 +12,18 @@ class StubRasterizer(Rasterizer):
...
         """
 
-        Arguments:
-            raster_size {Tuple[int, int]} -- Desired output image size
-            pixel_size {np.ndarray} -- Dimensions of one pixel in the real world
-            ego_center {np.ndarray} -- Center of ego in the image, [0.5,0.5] would be in the image center.
-            filter_agents_threshold {float} -- Value between 0 and 1 used to filter uncertain agent detections
+        Args:
+            render_context (RenderContext): Render Context
+            filter_agents_threshold (float): Value between 0 and 1 used to filter uncertain agent detections
```
Comment (stefaniespeichert) [inline l5kit/l5kit/rasterization/stub_rasterizer.py:21]: Can you maybe provide more info on render context? You use this quite a lot but  I haven't seen an explanation. Like saying it consists of bla and blub because blabla

File: l5kit/l5kit/sampling/agent_sampling.py [hunk 32]
```diff
@@ -113,9 +111,9 @@ def generate_agent_sample(
             )[0]
         except IndexError:
             raise ValueError(f" track_id {selected_track_id} not in frame or below threshold")
-        agent_centroid = agent["centroid"]
-        agent_yaw = float(agent["yaw"])
-        agent_extent = agent["extent"]
+        agent_centroid_m = agent["centroid"]
+        agent_yaw_rad = float(agent["yaw"])
+        agent_extent_m = agent["extent"]
         selected_agent = agent
```
Comment (stefaniespeichert) [inline l5kit/l5kit/sampling/agent_sampling.py:117]: what does _m stand for? might be unclear naming

**Review by stefaniespeichert — COMMENTED** [review]
I have one maybe major problem with this. From what I understand, kagglers now need to change their code to work with this new function, right? I've seen the changes in the sample ipynb for agent prediction but could you maybe make a small jupyter notebook where you explain these changes and walk them through it? 
Like just how to load the new rasterizers and how to transform and where this changes will influence code?

File: l5kit/l5kit/sampling/agent_sampling.py (hunk 32 above)
Comment (yye-lyft) [reply l5kit/l5kit/sampling/agent_sampling.py:117]: it stands for `meter`

Comment (lucabergamini) [reply l5kit/l5kit/sampling/agent_sampling.py:117]: metres :)

File: l5kit/l5kit/rasterization/stub_rasterizer.py (hunk 31 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/stub_rasterizer.py:21]: I'll do my best but I was not the one who added it actually :)

File: coords_systems.md [hunk 33]
```diff
@@ -46,9 +62,10 @@ One application of this is drawing trajectories on the image:
 data = dataset[0]  # this comes from either EgoDataset or AgentDataset
 
 im = dataset.rasterizer.to_rgb(data["image"].transpose(1, 2, 0))  # convert raster into rgb
-
+# transform points to world space
+positions_world = transform_points(data["target_positions"], data["world_from_agent"])
```
Comment (yye-lyft) [inline coords_systems.md:66]:
```suggestion
positions_in_world = transform_points(data["target_positions"], data["world_from_agent"])
```

**Comment by lucabergamini** [comment]
> Like just how to load the new rasterizers

This is not required as it's masked by `rasterizer_builder` :)



> maybe make a small jupyter notebook where you explain these changes and walk them through it?

They already have the examples and the doc files. I can add a cell into the visualisation notebook to make it even more clear though. We'll also have release notes with this :)

File: coords_systems.md [hunk 34]
```diff
@@ -46,9 +62,10 @@ One application of this is drawing trajectories on the image:
 data = dataset[0]  # this comes from either EgoDataset or AgentDataset
 
 im = dataset.rasterizer.to_rgb(data["image"].transpose(1, 2, 0))  # convert raster into rgb
-
+# transform points to world space
```
Comment (yye-lyft) [inline coords_systems.md:65]:
```suggestion
# transform points from agent local coordinate to world coordinate
```

File: coords_systems.md [hunk 35]
```diff
@@ -46,9 +62,10 @@ One application of this is drawing trajectories on the image:
...
+positions_world = transform_points(data["target_positions"], data["world_from_agent"])
 # transform from world meters into image pixels
-positions_pixels = transform_points(data["target_positions"] + data["centroid"][:2], data["world_to_image"])
```
Comment (yye-lyft) [inline coords_systems.md:51]: `# transform points from world coordinates into raster`

File: l5kit/l5kit/rasterization/semantic_rasterizer.py (hunk 30 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/semantic_rasterizer.py]: legacy, we can remove it

**Comment by yye-lyft** [comment]
Some comments, otherwise LGTM!

File: l5kit/l5kit/rasterization/satellite_rasterizer.py (hunk 27 above)
Comment (lucabergamini) [reply l5kit/l5kit/rasterization/satellite_rasterizer.py:70]: I've changed it to `meters` :) Although I prefer the British version (obviously)

**Review by lucabergamini — APPROVED** [review]
